enhancements/fail-forward-upgrades.md

```diff
@@ -0,0 +1,244 @@
+- The operator that initiated the upgrade, `operator v1`.
+- The operator that failed to install, `operator v2`.
+
+Lets consider what the user experience looks like depending on which version is sent to the resolver.
+
+#### Alternative 1a: Determine installabes based on the operator version that initiated the upgrade
+
+By providing the resolver with the `operator v1`, OLM would recalculate the list of installables using the set of operators that initiated the failed upgrade. OLM would only create a new InstallPlan once the catalog was updated to include a new upgrade path for the `operator v1`.
+
+##### Pros
+
+- The catalog curator/operator author must be aware of the failed installation, allowing them to address the issue directly by providing a new upgrade path.
+- Involving catalog curators/operator authors helps prevent unsafe/insane upgrades.
+- Upgrade graph maintenance relies tools/concepts embraced today.
+
+#### Cons
+
+- The catalog curator/operator author must manually address each failed installation.
+- Cluster admins cannot role forward until a new catalog is published.
+- Deviates significantly from OLM's upgrade behavior today. By the time `operator v2` has failed, most resources have been updated to include `operator v2` as an owner. This impacts OLMs upgrade logic in multiple places and introduces additional complexity to the solution.
+
+#### Alternative 1b: Determine installabes based on the operator version that failed to install
+
+By providing the resolver with the `operator v2`, OLM would attempt to immediately upgrade to the next version of the operator in the case of a failed installation. The catalog curator/operator author would not need to make any changes to the upgrade path of the operator.
+
+##### Pros
+
+- Catalog curators/operator authors do not need to recreate upgrade paths if they release a operator version that can't be installed.
+- Cluster admins can "fail forward" without consulting the catalog curator/operator author.
+- Doesn't deviate from OLM's existing upgrade logic today and therefor is relatively simple to implement.
+
+##### Cons
+
+- The solution does not guarantee any involvement from the catalog curator / operator author. Many customers may hit the issue before the team is notified.
+- Without involving the catalog curator/operator author there is increased likelihood of an unsafe upgrade. The operator version that failed to install might have been required due to a migration or some other critical reason.
+
+### Alternative 2: Address with RukPak APIs
+
+The rukpak APIs can provide this same toggle with fewer potential side effects, but does require more effort to implement. However, while the InstallPlan APIs will eventually be deprecated in favor of rukpak, this would take potentially another release cycle or more to get those changes in.
```

Comment:
Rukpak wouldn't provide this type of toggle out of the box, as the intent is to rely on higher level abstractions to handle these types of complex pivoting decisions. I imagine this type of failforward logic would be more likely to live in Deppy and the standalone resolver.

Reply:
Fair point. I'll change this to "Address with next gen apis".